### ISSUE TITLE
TypeScript/JavaScript TextMate syntax highlighting missing/inconsistent

### ISSUE THREAD

**Issue by minestarks** [mono/monodevelop#4144]
0. Build and run master as of today
1. File -> New -> Web -> TypeScript file
2. Copy and paste code below

```
/**
 * A function
 */
function foo(x: number, y: string) {
   return 3;
}

class A {
}

new A();

enum B {
    c,
    d,
    e
}

B.e

// Comment
function bar() {
    foo(3,"a").toExponential()
    return "asdf"
}
```

Result:
![image](https://user-images.githubusercontent.com/16928427/37358416-4c1369c4-26a8-11e8-9ed7-9be27cd37cf6.png)

Notice: 
- `function` keyword inconsistently highlighted in lines 4 and 22
- numeric literals (`3`) inconsistently highlighted in lines 5 and 23
- Line 21 Comment not highlighted
- Identifiers not highlighted at all

For comparison, VS Code:
![image](https://user-images.githubusercontent.com/16928427/37358440-58a78530-26a8-11e8-902f-a339ee48ca19.png)

> VS bug [#581503](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/581503)

**Comment by minestarks** [comment]
@Therzok FYI

**Comment by Therzok** [comment]
cc @mkrueger do we have outdated textmate highlighting for typescript?

otherwise, has the `StringBuilderCache` usage regressed somehow? I don't see how that could affect it right now, as we clear the string builder on both allocation and freeing, but re-reviewing that PR.

**Comment by minestarks** [comment]
In case it's helpful in tracking down the regression, this seems to have been reported in Oct 2017 in VSMac as well: https://developercommunity.visualstudio.com/content/problem/131494/getelementbyid-javascript-key-work-highlighting-an.html

**Comment by mkrueger** [comment]
Fixed in d17-razor branch. I added a work around for the issue - we're hitting here a difference in the regex engine behavior.

I hope we'll switch soon to the vs.net text mate regex engine.

**Comment by xamarin-release-manager** [comment]
> [sync] [VS-19] Comment by **xamarinc**



Fixed in version 8.6.0.802 (master)



Author: Mike Kr??ger  
 Commit: 474a97d694cc370f352217592b4270fb038e0048 (xamarin/monodevelop)
